server/eclipse-project/src/main/java/us/freeandfair/corla/auth/AbstractAuthentication.java

```diff
@@ -0,0 +1,237 @@
+    }
+
+    return result;
+  }
+  
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public County authenticatedCounty(final Request the_request) {
+    County result = null;
+    if (isAuthenticatedAs(the_request, AdministratorType.COUNTY,
+                    the_request.queryParams(AuthenticationInterface.USERNAME))) {
+      final Administrator admin = 
+          (Administrator) the_request.session().attribute(ADMIN);
+      if (admin != null) {
+        result = CountyQueries.forAdministrator(admin);
+      }
+    }
+    return result;
+  }
+  
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void traditionalDeauthenticate(final Request the_request,
+                                        final String the_username) {
+    the_request.session().removeAttribute(ADMIN);
+    Main.LOGGER.info("session is now traditionally deauthenticated");
+  }
+  
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void twoFactorDeauthenticate(final Request the_request,
+                                      final String the_username) {
+    the_request.session().removeAttribute(ADMIN);
+    Main.LOGGER.info("session is now second factor deauthenticated");
+  }
+  
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public boolean traditionalAuthenticated(final Request the_request,
+                                          final String the_username) {
+    final String auth_stage = 
+        the_request.session().attribute(AuthenticationInterface.AUTH_STAGE);
+    return auth_stage != null && 
+        (auth_stage.equals(AuthenticationInterface.SECOND_FACTOR) ||
+            auth_stage.equals(AuthenticationInterface.ADMIN));
+  }
+  
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public boolean secondFactorAuthenticated(final Request the_request,
+                                           final String the_username) {
+    final String auth_stage = 
+        the_request.session().attribute(AuthenticationInterface.AUTH_STAGE);
+    return auth_stage != null && 
+        auth_stage.equals(AuthenticationInterface.ADMIN);
+  }
+  
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void logoutAdministrator(final Request the_request,
+                                  final String the_username) {
+    Main.LOGGER.info("Logging out administrator `" + the_username + "'");
+    deauthenticate(the_request, the_username);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public boolean isAuthenticatedAs(final Request the_request,
+                                   final AdministratorType the_type,
+                                   final String the_username) {
+    boolean result = false;
+    final Object admin_attribute = the_request.session().attribute(ADMIN);
+    if (admin_attribute instanceof Administrator) {
+      final Administrator admin = (Administrator) admin_attribute;
+      result = admin.type() == the_type;
+      the_request.session().attribute(ADMIN, admin);
```

Comment:
I'm not sure why you're putting the object back in the session again here... you didn't remove it, so it should still be there, and you haven't changed it.

Reply:
That's a line straight from your implementation.  I was puzzled by it too.

Reply:
Well, let's get rid of it. I suspect it was a cut and paste error on my part that nobody caught in review, to be honest. :)